source/simulator/core.cc

```diff
@@ -2147,6 +2162,329 @@ namespace aspect
+                                        linearized_stokes_initial_guess,
+                                        current_linearization_point);
+
+                  compute_current_constraints ();
+                  assemblers.reset (new internal::Assembly::AssemblerLists<dim>());
+                  set_assemblers();
+                  assemble_stokes_system();
+
+                  this->last_pressure_normalization_adjustment = normalize_pressure(current_linearization_point);
+
+                  initial_newton_residual_velo = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+                  initial_newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  initial_newton_residual = std::sqrt(initial_newton_residual_velo * initial_newton_residual_velo + initial_newton_residual_pres * initial_newton_residual_pres);
+
+                  current_linearization_point.block(introspection.block_indices.velocities) = global_backup_linearization_point.block(introspection.block_indices.velocities);
+
+                  parameters.switch_initial_newton_residual = initial_newton_residual;
+                  parameters.newton_residual = initial_newton_residual;
+                  pcout << "   Initial Newton Stokes residual = " << initial_newton_residual << ", v = " << initial_newton_residual_velo << ", p = " << initial_newton_residual_pres << std::endl << std::endl;
+                }
+
+              assemble_newton_stokes_system = assemble_newton_stokes_matrix = true;
+
+              denormalize_pressure (this->last_pressure_normalization_adjustment,
+                                    linearized_stokes_initial_guess,
+                                    current_linearization_point);
+
+              compute_current_constraints ();
+
+              // the Stokes matrix depends on the viscosity. if the viscosity
+              // depends on other solution variables, then after we need to
+              // update the Stokes matrix in every time step and so need to set
+              // the following flag. if we change the Stokes matrix we also
+              // need to update the Stokes preconditioner.
+              rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+
+              assemble_stokes_system();
+
+              /**
+               * Eisenstat Walker method for determining the tolerance
+               */
+              if (nonlinear_iteration > 1)
+                {
+                  newton_residual_old = newton_residual;
+                  newton_residual_velo = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+                  newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+                  newton_residual = std::sqrt(newton_residual_velo * newton_residual_velo + newton_residual_pres * newton_residual_pres);
+
+                  if (!use_picard)
```

Comment:
Please move lines 2340-2375 into its own helper function as well. Also is the function *newton* specific or would it work with other iterative solver schemes as well? If so, please rename the variables to remove Newton specific names. The new function could have a function head like:
```
double Simulator<dim>::compute_linear_solver_tolerance(const double nonlinear_residual,
const double old_nonlinear_residual,
const double stokes_residual)
```
and the new code here could look like:
```
if (!use_picard)
  {
    parameters.linear_stokes_solver_tolerance = compute_linear_solver_tolerance(...);
    pcout << "The linear solver tolerance is set to ...;
  }
```